datafusion/core/src/physical_plan/aggregates/row_hash.rs

```diff
@@ -519,40 +556,100 @@ impl GroupedHashAggregateStream {
+        memory_delta.update(&mut self.reservation)
     }
 
-    /// Create an output RecordBatch with all group keys and accumulator states/values
-    fn create_batch_from_map(&mut self) -> Result<RecordBatch> {
+    /// Create an output RecordBatch with the group keys and
+    /// accumulator states/values specified in emit_to
+    fn create_batch_from_map(&mut self, emit_to: EmitTo) -> Result<RecordBatch> {
         if self.group_values.num_rows() == 0 {
-            let schema = self.schema.clone();
-            return Ok(RecordBatch::new_empty(schema));
+            return Ok(RecordBatch::new_empty(self.schema()));
         }
 
+        let output = self.build_output(emit_to)?;
+        self.remove_emitted(emit_to)?;
+        let batch = RecordBatch::try_new(self.schema(), output)?;
+        Ok(batch)
+    }
+
+    /// Creates output: `(group 1, group 2, ... agg 1, agg 2, ...)`
+    fn build_output(&mut self, emit_to: EmitTo) -> Result<Vec<ArrayRef>> {
         // First output rows are the groups
-        let groups_rows = self.group_values.iter();
-        let mut output: Vec<ArrayRef> = self.row_converter.convert_rows(groups_rows)?;
+        let mut output: Vec<ArrayRef> = match emit_to {
+            EmitTo::All => {
+                let groups_rows = self.group_values.iter();
+                self.row_converter.convert_rows(groups_rows)?
+            }
+            EmitTo::First(n) => {
+                let groups_rows = self.group_values.iter().take(n);
+                self.row_converter.convert_rows(groups_rows)?
+            }
+        };
 
-        // Next output each aggregate value, from the accumulators
+        // Next output each aggregate value
         for acc in self.accumulators.iter_mut() {
             match self.mode {
-                AggregateMode::Partial => output.extend(acc.state()?),
+                AggregateMode::Partial => output.extend(acc.state(emit_to)?),
                 AggregateMode::Final
                 | AggregateMode::FinalPartitioned
                 | AggregateMode::Single
-                | AggregateMode::SinglePartitioned => output.push(acc.evaluate()?),
+                | AggregateMode::SinglePartitioned => output.push(acc.evaluate(emit_to)?),
             }
         }
 
-        Ok(RecordBatch::try_new(self.schema.clone(), output)?)
+        Ok(output)
+    }
+
+    /// Removes the first `n` groups, adjusting all group_indices
+    /// appropriately
+    fn remove_emitted(&mut self, emit_to: EmitTo) -> Result<()> {
+        let mut memory_delta = MemoryDelta::new();
+        memory_delta.dec(self.state_size());
+
+        match emit_to {
+            EmitTo::All => {
+                // Eventually we may also want to clear the hash table here
+                //self.map.clear();
+            }
+            EmitTo::First(n) => {
+                // Clear out first n group keys by copying them to a new Rows.
+                // TODO file some ticket in arrow-rs to make this more efficent?
```

Comment:
I'm not sure this can be made efficient, but I also am not sure this aggregator needs to use the row format at all

Reply:
In this case, the group_values also need to support "remove the first N values" even if we removed the row format from the partial order state